director/template/template.go

```diff
@@ -172,25 +173,43 @@ func (l varsLookup) Get(name string) (interface{}, bool, error) {
 	}
 
 	if len(splitName) > 1 {
-		tokens := []patch.Token{patch.RootToken{}}
-
-		for _, token := range splitName[1:] {
-			tokens = append(tokens, patch.KeyToken{Key: token})
-		}
-
-		findOp := patch.FindOp{
-			Path: patch.NewPointer(tokens),
+		pointer, err := patch.NewPointerFromString("/" + strings.Join(splitName[1:], "/"))
+		if err != nil {
+			return nil, false, err
 		}
 
-		val, err = findOp.Apply(val)
+		val, found, err = processPatchPointer(pointer, val)
 		if err != nil {
-			return nil, false, err
+			tokens := []patch.Token{}
+			for _, token := range pointer.Tokens() {
```

Comment:
Replacing all index with key won't work for arbitrary nesting (e.g. if the path has an array index and a numeric key in it).

Looking at the short-circuit code at L170 I'm wondering if a token-by-token approach might be cleaner. Like your suggested go-patch PR that could allow us to reflect on the value for each token and choose the right token, instead of a blind try and retry.

Reply:
It turns out that numeric keys aren't a thing with go-patch.
It only looks up keys that are strings.
So that is not an issue, because it never supported it.

For example,

```bash
$ bosh int <(echo "person: ((names.1))") -l <(echo "names: {0: Bob, 1: Susie}")
Finding variable 'names.1':
  Expected to find a map key '1' for path '/1' (found map keys: '')

Exit code 1
```

This is with keys as numbers. It does not work.

When the keys are strings, just happen to be a number:

```bash
$ bosh int <(echo "person: ((names.1))") -l <(echo "names: {'0': Bob, '1': Susie}")
person: Susie

Succeeded
```

It works.

This is current behaviour of the bosh-cli, so this is maintained with the above code.

Reply:
I see, so the current state is:
- go-patch assumes all numbers are array indexes
- bosh assumes all numbers are string map keys
- the current PR tries the go-patch way, then the bosh way

I'd be interested in generalizing the solution to support mixing:
`bosh int <(echo "person: ((0.0))") -l <(echo "'0':['a']")`

But since this is all net new backwards compatible it can be a follow up PR.

Reply:
That becomes a combinatory issue then.
If you'd like that to work, then the request for `go-patch` to support `IndexToken` and `KeyToken` as union type would be better.